README
Gearman::Mesh

 thin wrappers around Gearman::XS modules

GEARMAND INSTALLATION

Installation will work without gearmand installed, but the distribution
will not be of much use.  To install gearmand:

CentOS
------

You'll need the EPEL repository installed.  After that:

 yum install gearmand

 # To start the service:
 service start gearmand

 # To enable at boot:
 chkconfig --level 245 gearmand on

OS X
----

Using Homebrew:

 brew install gearman

Using MacPorts (sudo may be required):

 port install gearmand

GEARMAN::MESH INSTALLATION

You can install Gearman::Mesh directly from GitHub with App::cpanminus:

 cpanm git@github.com:mikeraynham/Gearman-Mesh.git

Alternatively, if you have downloaded the distribution, run the following
commands:

 perl Build.PL
 ./Build
 ./Build test
 ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

 perldoc Gearman::Mesh
 perldoc Gearman::Mesh::Client
 perldoc Gearman::Mesh::Worker

LICENSE AND COPYRIGHT

Copyright (C) 2014 Mike Raynham, SpareRoom

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Gearman-Mesh

0.01    2015-08-15
        First version, released on an unsuspecting world.
0.02    2015-09-21
        Added automatic retries to worker and client.
0.03    2016-03-15
        Support for coalescing.
0.04    2017-02-21
        Fixed failing tests.
        Removed client retry.
        Improved long running worker test.
        Automatically start gearman daemon in tests.
0.05    2019-04-01
        Fixed failing tests.
0.06    2019-04-01
        Fixed failing tests. Used correct value this time.

t/lib/TestGearman.pm
package TestGearman;

use strict;
use warnings;

use Test::TCP ();
use File::Which ();
use Proc::Guard ();

sub new {
    my $class    = shift;
    my $gearmand = File::Which::which('gearmand');

    my %args = (
        'port'     => Test::TCP::empty_port(),
        'listen'   => '0.0.0.0',
        'log-file' => 'none',
    );

    my @args = map { sprintf('--%s=%s', $_, $args{$_}) } keys %args;
    my $proc = Proc::Guard->new(command => [ $gearmand, @args ]);

    Test::TCP::wait_port($args{port});

    return bless {
        host  => $args{listen},
        port  => $args{port},
        _proc => $proc
    }, $class;
}

sub host {$_[0]->{host}}
sub port {$_[0]->{port}}

1;


lib/Gearman/Mesh.pm
package Gearman::Mesh;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Carp qw(croak);
use Gearman::XS 0.16 qw(:constants);
use JSON::XS;

=head1 NAME

Gearman::Mesh - thin wrappers around Gearman::XS modules

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 DESCRIPTION

Gearman::Mesh provides thin wrappers for L<Gearman::XS::Client> and
L<Gearman::XS::Worker>.  The wrappers serialize and deserialize the
data that you pass to workers, and they make it a little easier to
define your workers.  For portability, Gearman::Mesh uses JSON for
data serialization, but it is possible to specify your own serialization
methods.

The Gearman::Mesh module doesn't provide any public methods.  It is a
base class for L<Gearman::Mesh::Client> and L<Gearman::XS::Worker>.

=cut

sub import {
    return unless shift eq 'Gearman::Mesh';
    return unless @_;

    my $pkg = caller;
    return unless $pkg =~ /^Gearman::Mesh::/;

    for my $sub (@_) {
        eval join("\n",
            qq`package $pkg;`,
            qq`sub $sub {shift->{_delegate}->$sub(\@_)}`,
        );
    }
}

sub new {
    my $class    = shift;
    my $delegate = shift;
    my %args     = @_;

    if (not exists $args{serialize_methods}) {
        $args{serialize_methods} ||= [
            \&JSON::XS::encode_json,
            \&JSON::XS::decode_json,
        ];
    }

    my $self = bless {
        _delegate => $delegate->new,
        %args,
    }, $class;

    if (exists $args{servers}) {
        $self->add_servers($args{servers}) == GEARMAN_SUCCESS
            || croak 'Failed to add servers';
    }

    return $self;
}

sub add_servers {
    my $self    = shift;
    my $servers = shift;

    if (ref $servers eq 'ARRAY') {
        $servers = join ',', @$servers;
    }
    elsif (ref $servers eq 'HASH') {
        $servers = join ',', map {$_ . ':' . $servers->{$_}}  keys %$servers;
    }

    $self->{_delegate}->add_servers($servers);
}

sub serialize {
    $_[0]->{serialize_methods}->[0]->($_[1]);
}

sub deserialize {
    $_[0]->{serialize_methods}->[1]->($_[1]);
}

=head1 SEE ALSO

L<< Gearman documentation|http://gearman.info/ >>

L<Gearman::XS> is a Perl front-end for the Gearman C library.  It provides
an almost raw interface to libgearman.  If you want full control over your
Gearman clients, workers, jobs and tasks, this is the module to use.

L<Gearman::Client>, L<Gearman::Task> and L<Gearman::Worker> are pure-Perl APIs
for Gearman.  They provide a simpler interface than L<Gearman::XS>.

L<AnyEvent::Gearman> provides asynchronous Gearman client and worker modules
for L<AnyEvent> applications.

=head1 AUTHOR

Mike Raynham, C<< <enquiries at mikeraynham.co.uk> >>

=head1 REPOSITORY

L<https://github.com/mikeraynham/Gearman-Mesh>

=head1 BUGS

Please report any bugs or feature requests to C<bug-gearman::mesh at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Gearman::Mesh>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Gearman::Mesh


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Gearman::Mesh>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Gearman::Mesh>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Gearman::Mesh>

=item * Search CPAN

L<http://search.cpan.org/dist/Gearman::Mesh/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 Mike Raynham, SpareRoom.co.uk.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Gearman::Mesh


lib/Gearman/Mesh/Client.pm
package Gearman::Mesh::Client;

use 5.006;
use strict;
use warnings FATAL => 'all';

use parent 'Gearman::Mesh';

use Gearman::Mesh qw(
    add_server
    remove_servers
    options
    set_options
    add_options
    remove_options
    timeout
    set_timeout
    echo
    run_tasks
    set_created_fn
    set_data_fn
    set_complete_fn
    set_fail_fn
    set_status_fn
    set_warning_fn
    error
    do_status
    job_status
    wait
    clear_fn
);

use Exporter 5.57 qw(import);
use Gearman::XS 0.16 qw(:constants);

our @EXPORT_OK   = @Gearman::XS::EXPORT_OK;
our %EXPORT_TAGS = %Gearman::XS::EXPORT_TAGS;

BEGIN {
    my @jobs = qw(
        do
        do_high
        do_low
        do_background
        do_high_background
        do_low_background
        add_task
        add_task_high
        add_task_low
        add_task_background
        add_task_high_background
        add_task_low_background
    );

    my $pkg = __PACKAGE__;

    for my $job (@jobs) {
        
        eval join("\n",
            qq`package $pkg;`,
             q`require Carp;`,
            qq`sub $job {`,
             q`    my $self     = shift;`,
             q`    my $task     = shift;`,
             q`    my $workload = shift;`,
             q`    my $unique   = shift;`,
            qq`    my \@ret = \$self->{_delegate}->$job(`,
             q`        $task,`,
             q`        $self->serialize([$workload]),`,
             q`        defined $unique ? $unique : (),`,
             q`    );`,
             q`    return if !defined $ret[0];`,
             q`    return if $ret[0] ne GEARMAN_SUCCESS;`,
             q`    return $ret[1];`,
             q`}`,
        );
    }
}

=head1 NAME

Gearman::Mesh::Client - A wrapper around Gearman::XS::Client

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';


=head1 SYNOPSIS

 use Gearman::Mesh::Client;

 my $client = Gearman::Mesh::Client->new(servers => [
     'gearman1:4730',
     'gearman2:4730',
 ]);

 $client->do_background(reverse => [qw(foo bar baz)]);

=head1 CONSTANTS

This module optionally re-exports the constants from L<Gearman::XS>:

 use Gearman::Mesh::Client qw(:constants);

=head1 CONSTRUCTOR

=head2 new

 my $worker = Gearman::Mesh::Worker(
    servers           => $servers?,
    serialize_methods => \@serialize_methods,
 );

=cut

sub new {
    shift->SUPER::new('Gearman::XS::Client', @_);
}

=head1 METHODS

=head2 do, do_high, do_low
=head2 do_background, do_high_background, do_low_background
=head2 add_task, add_task_high, add_task_low
=head2 add_task_background, add_task_high_background, add_task_low_background

 my $job_handle = $client->do($function_name => $workload, $unique?);
 my $task = $client->add_task($function_name => $workload, $unique?);

Each of these methods are wrapped in a method that serializes C<$workload>
before calling the equivalent L<Gearman::XS::Client> method.

Each method will retry up to ten times, using an exponential backoff algorithm
to determine the delay between retries.

Job methods (those whose names begin with C<do>) will return the job handle on
success.  Task methods (those whose name begin with C<add_task>) will return a
L<Gearman::XS::Task> object on success.  On failure, the methods will return
nothing.

The optional C<$unique> parameter is a string that uniquely identifies the job.
It is used by gearmand to coalesce jobs: if multiple clients request a job with
the same unique identifier, gearmand will perform the job only once and return
the result to all the clients that requested it.  Please note that this only
works for jobs that are in the same queue at the same time.

=head2 add_server, remove_servers, options, set_options, add_options
=head2 remove_options, timeout, set_timeout, echo, run_tasks, set_created_fn
=head2 set_data_fn, set_complete_fn, set_fail_fn, set_status_fn
=head2 set_warning_fn, error, do_status, job_status, wait, clear_fn

These methods pass the call through to a delegate L<Gearman::XS::Worker>
object.


=cut

=head1 AUTHOR

Mike Raynham, C<< <enquiries at mikeraynham.co.uk> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-gearman::mesh at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Gearman::Mesh>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Gearman::Mesh::Client


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Gearman::Mesh>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Gearman::Mesh>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Gearman::Mesh>

=item * Search CPAN

L<http://search.cpan.org/dist/Gearman::Mesh/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 Mike Raynham.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Gearman::Mesh::Client


lib/Gearman/Mesh/Worker.pm
package Gearman::Mesh::Worker;

use 5.006;
use strict;
use warnings;

use parent 'Gearman::Mesh';

use Gearman::Mesh qw(
    add_server
    remove_servers
    echo
    work
    grab_job
    error
    options
    set_options
    add_options
    remove_options
    timeout
    set_timeout
    register
    unregister
    unregister_all
    function_exist
    wait
    set_log_fn
);

use Scalar::Util qw(weaken refaddr);
use Exporter 5.57 qw(import);
use Gearman::XS 0.16 qw(:constants);
use Sub::Name;
 
our @EXPORT_OK   = @Gearman::XS::EXPORT_OK;
our %EXPORT_TAGS = %Gearman::XS::EXPORT_TAGS;

=head1 NAME

Gearman::Mesh::Worker - A wrapper around Gearman::XS::Worker

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

 use Gearman::Mesh::Worker qw(:constants);

 my $worker = Gearman::Mesh::Worker->new(servers => 'localhost:4730');

 $worker->add_function(reverse => sub {
     my $job      = shift;
     my $args     = shift;
     my $workload = shift;

     @result = map {scalar reverse} @$workload;
 });

 $worker->work;

=head1 CONSTANTS

This module optionally re-exports the constants from L<Gearman::XS>:

 use Gearman::Mesh::Worker qw(:constants);

=head1 CONSTRUCTOR

=head2 new

 my $worker = Gearman::Mesh::Worker(
    servers           => $servers?,
    serialize_methods => \@serialize_methods,
 );

=cut

sub new {
    shift->SUPER::new('Gearman::XS::Worker', @_);
}

=head1 METHODS

=head2 add_function

 add_function(
    $name => $coderef,
    $args,
    $timeout,
 );

=cut

sub add_function {
    my $self    = shift;
    my $name    = shift;
    my $coderef = shift;
    my $args    = shift;
    my $timeout = shift;

    $args    = '' unless defined $args;
    $timeout = 0  unless defined $timeout;

    $self->{_delegate}->add_function(
        $name,
        $timeout,
        subname($name, sub {
            my ($args)     = @{$self->deserialize($_[1])};
            my ($workload) = @{$self->deserialize($_[0]->workload)};

            $coderef->($_[0], $args, $workload);
        }),
        $self->serialize([$args]),
    );
}

=head2 add_functions

 add_functions(
     $name => $coderef,
     $name => $coderef,
     ...
 );

A simple wrapper around add_function() which allows multiple functions to
be added with a single call.  It is not possible to specify $args or
$timeout with this method.

=cut

sub add_functions {
    my $self      = shift;
    my %functions = @_;

    while (my ($name, $coderef) = each %functions) {
        $self->add_function($name, $coderef);
    }
}

=head2 work_loop

 $worker->work_loop;

Sets the worker to non-blocking mode with a time-out of 1 second.  The worker
is then placed in an event loop.  The loop will be terminated gracefully if
the process receives a SIG_TERM or SIG_INT.

The commonly shown alternative is:

 while (1) {
     $worker->work();
 }

This can cause jobs to terminate before completion if the worker process
receives a SIG_TERM or SIG_INT.  Using work_loop() allows jobs to finish
gracefully.

=cut

sub work_loop {
    my $self    = shift;
    my $worker  = $self->{_delegate};
    my $options = $worker->options;
    my $timeout = $worker->timeout;

    $worker->set_timeout(1_000);

    my $work_ok = sub {
        my $code = shift;
        return 1 if $code == GEARMAN_IO_WAIT
                 || $code == GEARMAN_NO_JOBS
                 || $code == GEARMAN_TIMEOUT;
        return;
    };

    {
        my $continue = 1;
        my $handler  = sub {$continue = 0};

        local $SIG{TERM} = $handler;
        local $SIG{INT}  = $handler;

        while ($continue) {
            my $work = $worker->work;
            next if $work == GEARMAN_SUCCESS;
            sleep 1, next if $work_ok->($work);

            warn $worker->error;
            sleep 5; 
        }
    }

    $worker->set_options($options);
    $worker->set_timeout($timeout);

    return 1;
}

=head2 add_server, remove_servers, echo, work, grab_job, error, options
=head2 set_options, add_options, remove_options, timeout, set_timeout
=head2 register, unregister, unregister_all, function_exist, wait, set_log_fn

These methods pass the call through to a delegate L<Gearman::XS::Worker>
object.

=head1 AUTHOR

Mike Raynham, C<< <enquiries at mikeraynham.co.uk> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-gearman::mesh at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Gearman::Mesh>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Gearman::Mesh::Worker


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Gearman::Mesh>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Gearman::Mesh>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Gearman::Mesh>

=item * Search CPAN

L<http://search.cpan.org/dist/Gearman::Mesh/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 Mike Raynham.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Gearman::Mesh::Worker


Build.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Gearman::Mesh',
    license             => 'artistic_2',
    dist_author         => q{Mike Raynham, SpareRoom <enquiries@mikeraynham.co.uk>},
    dist_version_from   => 'lib/Gearman/Mesh.pm',
    release_status      => 'unstable',
    configure_requires => {
        'Module::Build' => 0,
    },
    build_requires => {
        'File::Which' => 0,
        'Proc::Guard' => 0,
        'Test::Fatal' => 0,
        'Test::More' => 0,
        'Test::TCP' => 0,
    },
    requires => {
        'Gearman::XS' => 0.16,
        'JSON::XS'    => 0,
        'Sub::Name'   => 0,
    },
    add_to_cleanup     => [ 'Gearman-Mesh-*' ],
    create_makefile_pl => 'traditional',
    meta_merge => {
        resources => {
            repository => 'https://github.com/mikeraynham/Gearman-Mesh',
        }
    },
);

$builder->create_build_script();
